### PR TITLE
[WIP] Conversion to DB Client (PostgreSQL and MongoDB)

### PR DESCRIPTION
**Pull request by jeniawhite**
### Explain the changes
1. Conversion to DB Client that supports PostgreSQL and MongoDB
Signed-off-by: Evgeniy Belyi <jeniawhite92@gmail.com>

### Issues: Fixed #xxx / Gap #xxx
1. 

### Testing Instructions:
1.

### CODE REVIEW

File: src/util/db_client.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+const config = require('../../config');
+const postgres_client = require('./postgres_client');
+const mongo_client = require('./mongo_client');
+const USE_POSTGRESQL = process.env.USE_POSTGRESQL || config.USE_POSTGRESQL;
+
+/**
+ * @returns { nb.DBClient }
+ */
+function instance() {
+    return USE_POSTGRESQL ? postgres_client.instance() : mongo_client.instance();
```
[the remark this answers is not in the document]
Comment (jeniawhite) [reply src/util/db_client.js:14]: Alright, I totally agree. 
This was something that I've used temporarily until we decide on a convention.
You can see that I try to pull the flag from the ENV and only if there is none, I will try to access the config.js.
@guymguym 
Do you have any opposition to changing the config.js to an ENUM (like the DB_TYPE in the operator and ENV)?

[the remark this answers is not in the document]
Comment (guymguym) [reply src/util/db_client.js:14]: makes sense!

**Review by dannyzaken — APPROVED** [review]
just the one comment regarding the db_type.

File: src/util/http_utils.js [hunk 2]
```diff
@@ -154,7 +154,7 @@ function _parse_one_http_range(range_str) {
         if (second) {
             end = Number(second);
             if (!Number.isInteger(end)) throw_ranges_error(400);
-            if (first) end += 1; // end is inclusive in http ranges
+            end += 1; // end is inclusive in http ranges
```
Comment (guymguym) [inline src/util/http_utils.js:157]: This is a deep scan fix, the change is equivalent, but only concern is if we had some bug before. I can't see a problem with the current logic but this is a bit delicate.

File: src/server/bg_services/tier_ttf_worker.js [hunk 3]
```diff
@@ -126,10 +126,6 @@ class TieringTTFWorker {
                 default:
                     chunks_to_rebuild = 1;
             }
-
-
-            if (!chunks_to_rebuild) continue;
```
[the remark this answers is not in the document]
Comment (guymguym) [reply src/server/bg_services/tier_ttf_worker.js:131]: @jackyalbo Can you take a look - this change is due to a deep scan complaint that chunks_to_rebuild is never false. The code before seems to always set to non falsy value, but is it intended??

[the remark this answers is not in the document]
Comment (jackyalbo) [reply src/server/bg_services/tier_ttf_worker.js:131]: @guymguym looks like remains of old code. As I already filter out buckets the won't have any chunks_to_rebuild (too big TTF) so all buckets will have at least one chunk moving - so this check is indeed redundant.

[the remark this answers is not in the document]
Comment (guymguym) [reply src/server/bg_services/tier_ttf_worker.js:131]: Thanks @jackyalbo

File: src/endpoint/s3/s3_utils.js [hunk 4]
```diff
@@ -376,17 +376,17 @@ function parse_tagging_header(req) {
...
+    const tag_set = new Set();
     tagging_params.forEach((value, key) => {
         const tag = { key, value };
         if (!_is_valid_tag_values(tag)) throw new S3Error(S3Error.InvalidTag);
-        if (tag_map.has(tag.key)) throw new S3Error(S3Error.InvalidTag);
-        tag_map.set(tag.key);
-        tag_set.push(tag);
+        if (tag_set.has(tag.key)) throw new S3Error(S3Error.InvalidTag);
+        tag_set.add(tag.key);
+        tag_arr.push(tag);
```
Comment (guymguym) [inline src/endpoint/s3/s3_utils.js:386]: Instead of collecting both in a set and into an array, you can just add to the set and then `return Array.from(tag_set)`

**Review by guymguym — CHANGES_REQUESTED** [review]

File: src/rpc/rpc_message.js [hunk 5]
```diff
@@ -21,7 +21,7 @@ const RPC_VERSION_NUMBER = Buffer.from([
 // possible in order to ensure complaince on both sides of the RPC call.
 function clone_customizer(value) {
     // Mongo object ids are always serialized into strings and never deserialized back.
-    if (mongo_utils.is_object_id(value)) {
+    if (db_client.instance().is_object_id(value)) {
```
[the remark this answers is not in the document]
Comment (jeniawhite) [reply src/rpc/rpc_message.js]: Moved back to mongo_utils because of FE builds failing on the pg-native package (since this code requires Postgres).
Did not want to add the package to the FE because we will move from ObjectId to string and this doesn't make sense to include packages to FE (which currently supports MongoDB).
It will be removed on the move to string anyway.

File: src/test/unit_tests/test_postgres_client.js [hunk 6]
```diff
@@ -1,195 +1,195 @@
...
-        };
-        await test_table.insert_one(data1);
-        const find_res = await test_table.find({ name: 'test_projection' }, { projection: { name: 1, version: 1 } });
-        assert.deepEqual(find_res, [_.pick(data1, ['_id', 'name', 'version'])], 'the returned data should contain only id, name and version');
-    });
-
-
-});
\ No newline at end of file
+// const mocha = require('mocha');
```
[the remark this answers is not in the document]
Comment (jeniawhite) [reply src/test/unit_tests/test_postgres_client.js:4]: Uncommented and they aren't in index.js

File: src/test/unit_tests/test_migration_to_postgres.js [hunk 7]
```diff
@@ -1,83 +1,83 @@
 /* Copyright (C) 2016 NooBaa */
 'use strict';
 
-// const { PostgresClient } = require('../../util/postgres_client');
-const mongo_client = require('../../util/mongo_client');
-const { Pool } = require('pg');
-const _ = require('lodash');
+// // const { PostgresClient } = require('../../util/postgres_client');
```
[the remark this answers is not in the document]
Comment (jeniawhite) [reply src/test/unit_tests/test_migration_to_postgres.js:4]: Uncommented and they aren't in index.js

File: src/server/node_services/nodes_store.js [hunk 8]
```diff
@@ -168,14 +166,13 @@ class NodesStore {
      * @param {number} [limit]
      * @param {Object} [fields] 
      */
-    find_nodes(query, limit, fields) {
-        return this._nodes.col().find(query, { limit, projection: fields })
-            .toArray()
-            .then(nodes => this._validate_all(nodes, 'warn'));
+    async find_nodes(query, limit, fields) {
+        const nodes = await this._nodes.find(query, { limit, projection: fields });
```
[the remark this answers is not in the document]
Comment (jeniawhite) [reply src/server/node_services/nodes_store.js:170]: https://docs.mongodb.com/manual/tutorial/project-fields-from-query-results/
From reading the typescript of MongoDB I saw that fields is depriciated and we should use projection instead:
```
export interface FindOneOptions {
    projection?: object;
    /**
     * @deprecated Use options.projection instead
     */
    fields?: object;
```

[the remark this answers is not in the document]
Comment (guymguym) [reply src/server/node_services/nodes_store.js:170]: good to know. so did you check if there are remaining code using `fields`?

[the remark this answers is not in the document]
Comment (jeniawhite) [reply src/server/node_services/nodes_store.js:170]: Yes. Did found one place and changed it.
`/src/tools/node_ssh.js`

File: src/test/unit_tests/coretest.js [hunk 9]
```diff
@@ -87,6 +86,11 @@ function new_rpc_client() {
     return server_rpc.rpc.new_client(rpc_client.options);
 }
 
+function init_all_collections() {
+    /* eslint-disable global-require */
+    require('../../server/notifications/alerts_log_store').instance();
```
[the remark this answers is not in the document]
Comment (jeniawhite) [reply src/test/unit_tests/coretest.js:91]: Did not see any errors in logs regarding other collections.

File: src/server/analytic_services/history_data_store.js [hunk 10]
```diff
@@ -34,36 +33,33 @@ class HistoryDataStore {
...
-                    time_stamp: 1,
-                    version_snapshot: 1,
-                })
-                .toArray()
-            );
+    async get_system_version_history() {
+        return this._history.find({
+            history_type: 'VERSION'
+        }, {
+            time_stamp: 1,
```
Comment (guymguym) [inline src/server/analytic_services/history_data_store.js:60]: projection missing

**Review by guymguym — CHANGES_REQUESTED** [review]
Look for those find/findOne calls which forgot to send the `projection` property

File: src/server/system_services/bucket_server.js [hunk 11]
```diff
@@ -1552,15 +1552,11 @@ function _calc_metrics({
     if (bucket.quota) {
         let quota_precent = system_utils.get_bucket_quota_usage_percent(bucket, bucket.quota);
         info.quota = _.omit(bucket.quota, 'value');
-        let quota_free = size_utils.json_to_bigint(bucket.quota.value).minus(size_utils.json_to_bigint(objects_aggregate.size));
```
[the remark this answers is not in the document]
Comment (guymguym) [reply src/server/system_services/bucket_server.js:1555]: @nimrod-becker ping

File: src/server/analytic_services/history_data_store.js (hunk 10 above)
Comment (jeniawhite) [reply src/server/analytic_services/history_data_store.js]: Somehow missed it thanks.
Browsed all files with .find( and .findOne(
Did not find any occurrences.

**Review by guymguym — APPROVED** [review]
Looks OK, but this should be tested installed and tested manually with mongodb to sanitize.

**Review by guymguym — CHANGES_REQUESTED** [review]
@jeniawhite Can you squash?

File: src/server/analytic_services/activity_log_store.js [hunk 12]
```diff
@@ -36,26 +35,25 @@ class ActivityLogStore {
...
-        if (!activity_log._id) {
-            activity_log._id = this.make_activity_log_id();
-        }
-        return P.resolve()
-            .then(() => this._activitylogs.validate(activity_log))
-            .then(() => this._activitylogs.col().insertOne(activity_log))
-            .catch(err => mongo_utils.check_duplicate_key_conflict(err, 'audit_log'))
-            .return(activity_log);
+    async create(activity_log) {
+        return P.resolve().then(async () => {
```
Comment (guymguym) [inline src/server/analytic_services/activity_log_store.js:39]: This would not work. Async function will convert anything to a native Promise, so doesn't matter what you return it won't be received by the caller as a Bluebird promise.
```javascript
$ node
Welcome to Node.js v12.18.4.
Type ".help" for more information.
> B = require('bluebird')
> f = async () => B.resolve(2)
[AsyncFunction: f]
> p = f()
Promise { <pending> }
> p instanceof B
false
> p instanceof Promise
true
```

File: src/server/system_services/stats_aggregator.js [hunk 13]
```diff
@@ -420,7 +420,7 @@ async function _partial_buckets_info(req) {
 
     try {
         for (const bucket of system_store.data.buckets) {
-            if (String(bucket.system._id) !== String(req.system._id)) return;
+            if (String(bucket.system._id) !== String(req.system._id)) continue;
```
Comment (guymguym) [inline src/server/system_services/stats_aggregator.js:423]: Is this fix related to this PR?

Comment (jeniawhite) [reply src/server/system_services/stats_aggregator.js:423]: The fix is not related to the PR, I just did that since it crashed the partial metrics which means Prometheus metrics when we have namespace buckets.

File: src/server/analytic_services/activity_log_store.js (hunk 12 above)
Comment (jeniawhite) [reply src/server/analytic_services/activity_log_store.js:39]: Missed removing async from `activity_log_store` and `history_data_store`.
Thank you.

**Review by guymguym — APPROVED** [review]
🏆
